### PR TITLE
add pvdaq reference sites, fetch, config

### LINKED ISSUE #397: PVDAQ sites for reference data
We need to add some of the PVDAQ sites to the reference data set. The data does not need to be available in real time - anytime since April 2019 (the start of the reference forecasts archive) would be good enough.

https://developer.nrel.gov/docs/solar/pvdaq-v3/

https://maps.nrel.gov/pvdaq/?aL=su70DP%255Bv%255D%3Dt&bL=clight&cE=0&lR=0&mC=40.212440718286466%2C-91.58203125&zL=4

There's a stale pvlib python PR for the PVDAQ api. Maybe start there.

https://github.com/pvlib/pvlib-python/pull/664

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #397  .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [ ] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [ ] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Additional To do: 
 - [x] make fetch wrapper functions. Need to reindex to fixed frequency (most sites report at only quasi-regular intervals). Also need to localize data in fetch wrapper because PVDAQ apparently returns local time (no DST), but without the tz offset. I pulled tz offset data from google's timezone API.
 - [x] add pv modeling parameters somewhere
 - [x] add sites to the reference forecast configuration too - not sure how that's done.
 - [x] set up SFA API key
 - [ ] add sites to google map accessible at https://solarforecastarbiter.org/referencedata/

`get_pvdaq_data` function is based on https://github.com/pvlib/pvlib-python/pull/664

It takes about 3 minutes to pull all of the 2020 data.

### CODE REVIEW

**Comment by wholmgren** [comment]
@alorenzo175 @lboeman I found this existing code for conforming to an index:

https://github.com/SolarArbiter/solarforecastarbiter-core/blob/642fad058d9f9618e0699c0064f5d14994a65c5f/solarforecastarbiter/io/reference_observations/common.py#L309-L315

Can we change that to 

```python
    data_resampled = data.resample(observation.interval_length).first()
    new_index = pd.date_range(start=data.index[0], end=data.index[-1],
                              freq=observation.interval_length)
    data = data.reindex(new_index)
```

**Comment by alorenzo175** [comment]
Hmm, and if interval label is ending? Seems like ``.last()`` is more appropriate. And if ``interval_label == 'mean'`` seems like a mean would be appropriate?

**Comment by wholmgren** [comment]
I was originally just trying to fix a problem with timestamps that need to be rounded to be on the right interval. Here are a couple of examples:

<table class="dataframe" border="1">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ac_power</th>
      <th>dc_power</th>
    </tr>
    <tr>
      <th>Date-Time</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-03-12 13:30:00</th>
      <td>61000.0</td>
      <td>64600.0</td>
    </tr>
    <tr>
      <th>2020-03-12 13:35:00</th>
      <td>101700.0</td>
      <td>107700.0</td>
    </tr>
    <tr>
      <th>2020-03-12 13:40:00</th>
      <td>60300.0</td>
      <td>63800.0</td>
    </tr>
    <tr>
      <th>2020-03-12 13:45:00</th>
      <td>57300.0</td>
      <td>60800.0</td>
    </tr>
    <tr>
      <th>2020-03-12 13:49:58</th>
      <td>58400.0</td>
      <td>61800.0</td>
    </tr>
    <tr>
      <th>2020-03-12 13:50:00</th>
      <td>58400.0</td>
      <td>62100.0</td>
    </tr>
    <tr>
      <th>2020-03-12 13:55:00</th>
      <td>67200.0</td>
      <td>71100.0</td>
    </tr>
  </tbody>
</table>
</div></div>

and 

<table class="dataframe" border="1">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ac_power</th>
    </tr>
    <tr>
      <th>Date-Time</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2020-01-04 12:00:00</th>
      <td>1057</td>
    </tr>
    <tr>
      <th>2020-01-04 12:15:01</th>
      <td>1685</td>
    </tr>
    <tr>
      <th>2020-01-04 12:30:00</th>
      <td>1982</td>
    </tr>
    <tr>
      <th>2020-01-04 12:45:00</th>
      <td>2891</td>
    </tr>
    <tr>
      <th>2020-01-04 13:00:00</th>
      <td>2428</td>
    </tr>
    <tr>
      <th>2020-01-04 13:15:00</th>
      <td>2285</td>
    </tr>
    <tr>
      <th>2020-01-04 13:30:00</th>
      <td>2169</td>
    </tr>
    <tr>
      <th>2020-01-04 13:45:00</th>
      <td>2075</td>
    </tr>
    <tr>
      <th>2020-01-04 14:00:00</th>
      <td>1418</td>
    </tr>
    <tr>
      <th>2020-01-04 14:15:00</th>
      <td>1829</td>
    </tr>
    <tr>
      <th>2020-01-04 14:30:01</th>
      <td>1181</td>
    </tr>
    <tr>
      <th>2020-01-04 14:45:00</th>
      <td>2127</td>
    </tr>
  </tbody>
</table>

In the first case we could average or we could just discard. In the second case we want to round. If discarding is ok, then we can use `resample().first()` regardless of the label. 

But there are also two sites that report 15s data that we'll need to resample to 1 minute intervals (unless we want to support non integer interval_lengths in the API).

**Comment by alorenzo175** [comment]
Dumping data is ifne with me

**Comment by wholmgren** [comment]
I'd like some feedback on the proposed pattern before implementing/fixing tests.

`pvdaq_reference_sites.json` is a dict with keys `'sites'` and `'observations`'. The `sites` are all of the pvdaq sites, and the observations are a flat list of all of the pvdaq observations. I don't have a strong opinion about this vs. a nested structure. The code I'm parsing a handful of ugly python dicts and dataframes into `Site` and `Observation` objects. From there it's simple to change the organization of the json file if people prefer something else. I know @lboeman was thinking of something similar for SRML PV sites.

`common._prepare_data_to_post` gets a new `resample_how` argument that can be `None` or a pandas resampler method. For PVDAQ, this lets us drop data or resample sub-minute data to 1 minute data. `common.post_observation_data` pulls this from `observation.extra_parameters`, with default `None`. I could put this kind of logic in `reference_observations.pvdaq.fetch` if I assign the resample rules to the `Site` instead of the `Observation`. While that might be consistent with the existing expectations of all reference site data, I think we need to be more flexible in the future. I also just think it makes more sense to put it in the observation metadata.

**Comment by wholmgren** [comment]
It's now working with the `solararbiter referencedata init` and `update` commands. 

See this site for an example: https://dev-dashboard.solarforecastarbiter.org/observations/bfdb2b98-923f-11ea-bfa1-0a580a82013d?start=2020-04-01T22%3A00Z&end=2020-05-09T22%3A00Z

Observations won't `update` at the other sites because I first created them with not quite the correct `extra_parameters` (it was a dict instead of string-ified dict). I deleted the site/observation/forecasts at just the linked site above (major pain - separate issue) and then recreated it with extra parameters that are strings, and the values post was successful.

**Comment by wholmgren** [comment]
Looks like everything is working. There are a couple of sites on the dev dashboard that have the wrong AC/DC capacity since I hadn't yet fixed the W-->MW conversion when they were created, but they would be recreated correctly following a database wipe. It takes about 20 minutes to pull and post all of the 2020 data. About 75% of the time is in the post.

File: solarforecastarbiter/io/fetch/pvdaq.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        year = int(year)
+    except TypeError:
+        year = [int(yr) for yr in year]
+    else:
+        year = [year]
+
+    # Each year must queries separately, so iterate over the years and
+    # generate a list of dataframes.
+    # Consider putting this loop in its own private function with
+    # try / except / try again pattern for network issues and NREL API
```
Comment (alorenzo175) [inline solarforecastarbiter/io/fetch/pvdaq.py:87]: will we see any issues from this?

File: solarforecastarbiter/io/reference_observations/common.py [hunk 2]
```diff
@@ -291,21 +296,26 @@ def update_site_observations(api, fetch_func, site, observations,
         post_observation_data(api, obs, data_in_range, start, end)
 
 
-def _prepare_data_to_post(data, variable, observation, start, end):
+def _prepare_data_to_post(data, variable, observation, start, end,
+                          resample_how):
```
Comment (alorenzo175) [inline solarforecastarbiter/io/reference_observations/common.py:300]:
```suggestion
                          resample_how=None):
```

Comment (alorenzo175) [reply solarforecastarbiter/io/reference_observations/common.py:300]: and avoid having to add it to a bunch of tests

File: solarforecastarbiter/io/fetch/pvdaq.py (hunk 1 above)
Comment (wholmgren) [reply solarforecastarbiter/io/fetch/pvdaq.py:87]: The pvlib CI struggled with a different NREL API but I haven't run into any with the pvdaq API. They have a 1000 requests per hour limit but we are no where close to that. Let's see if it's a problem in the rc cycle.

**Comment by wholmgren** [comment]
@alorenzo175 tests pass, coverage is almost 100%, so I think we're close. The mocks grew more aggressive as the hour grew later, but I think they're consistent with other modules.

**Pull request merged**